### PR TITLE
fix: fix net stubbing flake

### LINKED ISSUE #20397: Intercepted requests can bleed over into the next test when the test ends quickly after interception
### Current behavior

When you intercept a request and then immediately move on to the next test, the response will occasionally not make it back to the client which will cause the browser to trigger an automatic retry. This can cause a subsequent test to pick up the call if it is also intercepting the same route. This can cause duplicate interception or an interception when one is explicitly being tested not to occur. An example of the second scenario can be seen below from [this code](https://github.com/cypress-io/cypress/blob/05f9ad163ec4ab13dc296b6ec184d28b5cea6670/packages/driver/cypress/integration/commands/net_stubbing_spec.ts#L3237-L3255):

image

Note that the request is flagged as "other" since the prerequest didn't happen in the current test. The code that does that is [here](https://github.com/cypress-io/cypress/blob/develop/packages/driver/src/cypress/proxy-logging.ts#L305-L315)

### Desired behavior

A request in one test should not be intercepted by a subsequent test.

### Test code to reproduce

This code is an example of what causes the problem:

https://github.com/cypress-io/cypress/blob/05f9ad163ec4ab13dc296b6ec184d28b5cea6670/packages/driver/cypress/integration/commands/net_stubbing_spec.ts#L3237-L3255

It is a race condition however so it will not occur consistently.

### Cypress Version

9

### Other

_No response_

### PR DESCRIPTION
**Pull request by ryanthemanuel**
- Closes #20397 

### User facing changelog

Intercepted requests will no longer bleed over to subsequent tests due to automatic browser retries.

### Additional details

On certain errors (like empty responses) the browser will retry. This behavior is specified in the W3C spec and can even happen with posts: https://www.w3.org/Protocols/rfc2616/rfc2616-sec8.html Destroying the socket is exactly what we do here: https://github.com/cypress-io/cypress/blob/c3eb6350fb0c5cee15da18ae766390c6796794dc/packages/net-stubbing/lib/server/state.ts#L9-L23

When ending the response rather than just destroying it, the retries do not occur which is wha want; however, we might cause unexpected behavior in the AUT by ending responses like that and so it might be good to only end/destroy when absolutely necessary, i.e. if a callback to the browser is unnecessary, like with a static response or spy we don't necessarily need to end/destroy. In the solution on this PR, we only end and destroy the response if we are not dealing with a spy or a static response and we have a subscription we are waiting on that we haven't fulfilled.

### PR Tasks

- [X] Have tests been added/updated?
- [na] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [na] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/net-stubbing/lib/server/state.ts [hunk 1]
```diff
@@ -1,24 +1,60 @@
...
+
+        if (shouldDestroyResponse) {
+          res.removeAllListeners('finish')
+          res.removeAllListeners('error')
+
+          return new Promise<void>((resolve) => {
+            res.once('finish', () => {
+              res.destroy()
+              delete this.requests[request.id]
+              resolve()
```
Comment (flotwig) [inline packages/net-stubbing/lib/server/state.ts:41]: do we actually need to wait for the request to finish to `resolve`? once it's `end`ed, why do we care?

File: packages/net-stubbing/lib/server/state.ts [hunk 2]
```diff
@@ -1,24 +1,60 @@
...
+            })
+
+            res.once('error', () => {
```
Comment (flotwig) [inline packages/net-stubbing/lib/server/state.ts:44]: do we need this and `.once('finish')`? I don't know why `.end` would sometimes cause `'error'` and sometimes cause `'finish'`

File: packages/net-stubbing/lib/server/state.ts [hunk 3]
```diff
@@ -1,24 +1,60 @@
...
+
+            res.end()
+          })
+        }
+
+        return Promise.resolve()
+      }))
 
       this.pendingEventHandlers = {}
-      this.requests = {}
```
Comment (flotwig) [inline packages/net-stubbing/lib/server/state.ts:21]: I see you've switched from using `this.requests = {}` to using `delete this.requests[requestId]`, but is there a reason to change this? is existing code referencing `this.requests[requestId]` and causing errors?

I'd argue that we don't want to get rid of this line - `InterceptedRequest` objects contain references to `BackendRoutes`, and we're trying to prevent leaking these route objects between tests to eliminate race conditions, so it seems like this object should be totally empty between tests, right?

File: packages/net-stubbing/lib/server/state.ts (hunk 2 above)
Comment (ryanthemanuel) [reply packages/net-stubbing/lib/server/state.ts:44]: Good call on this. I think I was trying to cover all of the bases here and make sure we never had an unfinished promise, but I never ran across this scenario in my testing. It is likely fine to remove.

File: packages/net-stubbing/lib/server/state.ts [hunk 4]
```diff
@@ -1,24 +1,60 @@
...
-        res.on('error', noop)
-        res.destroy()
-      }
+        if (res.destroyed) {
+          delete this.requests[request.id]
+
+          return Promise.resolve()
+        }
+
+        let shouldDestroyResponse = false
```
Comment (BlueWinds) [inline packages/net-stubbing/lib/server/state.ts:21]: So being unfamiliar with this code, it's rather hard to follow what's going on and why. Can you comment this section a bit more. explaining what's going on and why?

File: packages/net-stubbing/lib/server/state.ts [hunk 5]
```diff
@@ -1,24 +1,60 @@
...
+
+        request.subscriptionsByRoute.forEach((subscriptionByRoute) => {
```
Comment (BlueWinds) [inline packages/net-stubbing/lib/server/state.ts:23]: Can maybe get some more legibility by replacing the `forEach` loops with more functional style `.some()` calls.

```
const shouldAwaitSubscription = (sub) => (
  sub.await
  && !sub.skip
  && (responseEvents.includes(sub.eventName) || !inResponsePhase)
)

const shouldDestroyResponse = request.subscriptionsByRoute.some((subs) => (
  !subs.immediateStaticResponse
  && subs.subscriptions.some(shouldAwaitSubscription)
})
```

**Review by BlueWinds — CHANGES_REQUESTED** [review]

**Comment by emilyrohrbough** [comment]
Closing as stale. We can re-open when this is ready to go.
